libgnucash/engine/Account.cpp

```diff
@@ -3150,49 +3152,22 @@ xaccAccountGetName (const Account *acc)
 gchar *
 gnc_account_get_full_name(const Account *account)
 {
-    AccountPrivate *priv;
-    const Account *a;
-    char *fullname;
-    const gchar **names;
-    int level;
-
-    /* So much for hardening the API. Too many callers to this function don't
-     * bother to check if they have a non-NULL pointer before calling. */
-    if (NULL == account)
-        return g_strdup("");
-
-    /* errors */
-    g_return_val_if_fail(GNC_IS_ACCOUNT(account), g_strdup(""));
-
-    /* optimizations */
-    priv = GET_PRIVATE(account);
-    if (!priv->parent)
-        return g_strdup("");
+    std::string fullname{""};
+    std::string sep{account_separator};
 
-    /* Figure out how much space is needed by counting the nodes up to
-     * the root. */
-    level = 0;
-    for (a = account; a; a = priv->parent)
-    {
-        priv = GET_PRIVATE(a);
-        level++;
-    }
+    if (!account)
+        return g_strdup ("");
 
-    /* Get all the pointers in the right order. The root node "entry"
-     * becomes the terminating NULL pointer for the array of strings. */
-    names = (const gchar **)g_malloc(level * sizeof(gchar *));
-    names[--level] = NULL;
-    for (a = account; level > 0; a = priv->parent)
+    for (auto acc = (Account*)account; !gnc_account_is_root (acc);
+         acc = gnc_account_get_parent (acc))
     {
-        priv = GET_PRIVATE(a);
-        names[--level] = priv->accountName;
+        std::string name = xaccAccountGetName (acc);
+        if (acc == account)
+            fullname = name;
+        else
+            fullname = name + sep + fullname;
```

Comment:
Another pessimization: Allocations are expensive and you've got three for every step.